expr/tagquery/expression.go

```diff
@@ -45,134 +37,336 @@ func (e Expressions) Strings() []string {
+	// combinations can come to a conclusive decision without looking at all indexes and some others can't.
+	// if an expression has evaluated a metric against all indexes and has not come to a conclusive
+	// decision, then the default decision gets applied.
+	//
+	// Example
+	// metric1 has tags ["name=a.b.c", "some=value"] in the metric tag index, we evaluate the expression
+	// "anothertag!=value":
+	// 1) expression looks at the metric tag index and it sees that metric1 does not have a tag "anothertag"
+	//    with the value "value", but at this point it doesn't know if another index that will be looked
+	//    at later does, so it returns the decision "none".
+	// 2) expression now looks at index2 and sees again that metric1 does not have the tag and value
+	//    it is looking for, so it returns "none" again.
+	// 3) the expression execution sees that there are no more indexes left, so it applies the default
+	//    decision for the operator != which is "pass", meaning the expression "anothertag!=value" has
+	//    not filtered the metric metric1 out of the result set.
+	//
+	// metric2 has tags ["name=a.b.c", "anothertag=value"] according to the metric tag index and it has
+	// no meta tags, we still evaluate the same expression:
+	// 1) expression looks at metric tag index and see metric2 has tag "anothertag" with value "value".
+	//    it directly comes to a conclusive decision that this metric needs to be filtered out of the
+	//    result set and returns the filter decision "fail".
+	//
+	// metric3 has tags ["name=aaa", "abc=cba"] according to the metric tag index and there is a meta
+	// record assigning the tag "anothertag=value" to metrics matching that query expression "abc=cba".
+	// 1) expression looks at metric3 and sees it does not have the tag & value it's looking for, so
+	//    it returns the filter decision "none" because it cannot know for sure whether another index
+	//    will assign "anothertag=value" to metric3.
+	// 2) expression looks at the meta tag index and it sees that there are meta records matching the
+	//    tag "anothertag" and the value "value", so it retrieves the according filter functions of
+	//    of these meta records and passes metric3's tag set into them.
+	// 3) the filter function of the meta record for the query set "abc=cba" returns true, indicating
+	//    that its meta tag gets applied to metric3.
+	// 4) based on that the tag expression comes to the decision that metric3 should not be part of
+	//    final result set, so it returns "fail".
+	GetDefaultDecision() FilterDecision
 
-	equalPos = strings.Index(expression, "=")
-	if equalPos < 0 {
-		return res, fmt.Errorf("Missing equal sign: %s", expression)
-	}
+	// GetKey returns tag to who's values this expression get's applied if it operates on the value
+	// (OperatorsOnTag returns "false")
+	// example:
+	// in the expression "tag1=value" GetKey() would return "tag1" and OperatesOnTag() returns "false"
```

Comment:
this is confusing. i checked all the implementations and this always returns the tag key, wether or not OperatesOnTag is true or false.

Reply:
true, i think this comment is old and out of date. I'll update it